src/meta/src/hummock/hummock_manager.rs

```diff
@@ -1196,180 +1121,36 @@ where
-    }
-
-    #[named]
-    pub async fn get_ssts_to_delete(
-        &self,
-        version_id: HummockVersionId,
-    ) -> Result<Vec<HummockSSTableId>> {
-        let versioning_guard = read_lock!(self, versioning).await;
-        let _timer = start_measure_real_process_timer!(self);
-        Ok(versioning_guard
-            .stale_sstables
-            .get(&version_id)
-            .map(|s| s.id.clone())
-            .unwrap_or_default())
-    }
-
-    #[named]
-    pub async fn delete_will_not_be_used_ssts(
-        &self,
-        version_id: HummockVersionId,
-        ssts_in_use: &HashSet<HummockSSTableId>,
-    ) -> Result<()> {
+    pub async fn proceed_version_checkpoint(&self) -> Result<()> {
         let mut versioning_guard = write_lock!(self, versioning).await;
-        let _timer = start_measure_real_process_timer!(self);
-        let versioning = versioning_guard.deref_mut();
-        let mut stale_sstables = BTreeMapTransaction::new(&mut versioning.stale_sstables);
-        let mut sstable_id_infos = BTreeMapTransaction::new(&mut versioning.sstable_id_infos);
-        if let Some(mut ssts_to_delete) = stale_sstables.get_mut(version_id) {
-            // Delete sstables that are stale in the view of `version_id` Version, and
-            // are Not referred by any other older Version.
-            // No newer version would use any stale sstables in the current Version.
-            let num_ssts_to_delete = ssts_to_delete.id.len();
-            for idx in (0..num_ssts_to_delete).rev() {
-                let sst_id = ssts_to_delete.id[idx];
-                if !ssts_in_use.contains(&sst_id) && let Some(mut sst_id_info) = sstable_id_infos.get_mut(sst_id) {
-                    sst_id_info.meta_delete_timestamp = sstable_id_info::get_timestamp_now();
-                    // We don't want to repetitively set the delete timestamp of these that have been set,
-                    // so we remove these ones.
-                    ssts_to_delete.id.swap_remove(idx);
-                }
-            }
+        let min_pinned_version_id = versioning_guard.min_pinned_version_id();
+        if min_pinned_version_id <= versioning_guard.checkpoint_version.id {
+            return Ok(());
         }
-
-        commit_multi_var!(self, None, stale_sstables, sstable_id_infos)?;
-
-        #[cfg(test)]
+        let versioning = versioning_guard.deref_mut();
+        let mut checkpoint = VarTransaction::new(&mut versioning.checkpoint_version);
+        let old_checkpoint_id = checkpoint.id;
+        let new_checkpoint_id = min_pinned_version_id;
+        for (_, version_delta) in versioning
+            .hummock_version_deltas
+            .range((Excluded(old_checkpoint_id), Included(new_checkpoint_id)))
         {
-            drop(versioning_guard);
-            self.check_state_consistency().await;
+            checkpoint.apply_version_delta(version_delta);
         }
-
+        commit_multi_var!(self, None, checkpoint)?;
```

Comment:
if min_pinned_version_id == u64::MAX, it's possible that we do empty commit here. we can avoid this.

Reply:
fixed